gslib/wildcard_iterator.py

```diff
@@ -649,12 +651,12 @@ def __iter__(self, bucket_listing_fields=None):
     for filepath in filepaths:
       expanded_url = StorageUrlFromString(filepath)
       try:
-        if self.exclude_pattern:
-          base_url_str, _ = os.path.split(wildcard)
+        if self.exclude_tuple is not None:
+          (base_url_str, exclude_pattern) = self.exclude_tuple
```

Comment:
This works , but how about doing what I was suggesting here https://github.com/GoogleCloudPlatform/gsutil/pull/1602#discussion_r995250227. It will let us avoid all this logic involving file path slicing. This means, instead of passing the exclude_tuple, you can just pass the complete exclude pattern. i.e `os.path.join(base_url_str, exclude_pattern) if exclude_pattern else None` and then you will just have to do
```
if self.exclude_pattern is not None:
  if self.exclude_pattern.match(filepath):
    # log message
    continue
```

This also helps avoid passing in a tuple when a single string would suffice.

Reply:
We can do `exclude_pattern.pattern` to get the string back. I tried pulling your PR and making local changes, and it works. Pasting the relevant diff here.

```
-        expanded_exclude_pattern=re.compile(os.path.join(base_url_str, cls.exclude_pattern.pattern)),
+        exclude_tuple=(base_url_str, cls.exclude_pattern),

...

       try:
-        if self.expanded_exclude_pattern is not None:
-          if self.expanded_exclude_pattern.match(str(expanded_url)):
+        if self.exclude_tuple is not None:
+          (base_url_str, exclude_pattern) = self.exclude_tuple
+          str_to_check = filepath[len(base_url_str):]
+          if str_to_check.startswith(self.wildcard_url.delim):
+            str_to_check = str_to_check[1:]
+          if exclude_pattern.match(str_to_check):
             if self.logger:
               self.logger.info('Skipping excluded path %s...', filepath)
             continue

...

@@ -807,7 +811,7 @@ def CreateWildcardIterator(url_str,
                            gsutil_api,
                            all_versions=False,
                            project_id=None,
-                           expanded_exclude_pattern=None,
+                           exclude_tuple=(None, None),
```



The reason I feel like doing that is because the notion of a `base_url_str` is very rsync specific as rsync deals with a single base dir. What if we have source pattern that resolves to multiple dirs, so my suggestion is to do is try to avoid having rsync specific logic in wildcard iteartor. If you think of wildcard iterator on its own, what should we pass in as `base_url_str` is not clear. Let's say we want to use it for `cp` command in the future, then what should we pass? Note that the logic relies on the fact that the `base_url_str` should not have any wildcard, this is possible only in `rsync` since the base is already expanded. If we simply have a contract that says `send a regex pattern that should be used for filtering the listing`, then it's up to the caller on what pattern to pass down and hence the suggestion. For rsync it might look like`base_url_str + exclude_pattern`, for cp it might look like something else, and wildcard iterator should not have to worry about it. It will just use the pattern to filter out the results.

Regarding matching the logic in rsync, the current logic in rsync is based off of certain assumptions that rsync already has the base wildcard expanded and it only works on a single dir. Additionally it also does not call wildcard iterator for non-recursive cases (it lists the directory directly using `_LocalDirIterator`), so for that reason,  I feel we shouldn't try to match that logic in wildcard iterator.

Reply:
Thanks for the in-depth explanation... I just committed a version of your suggestion.

Reply:
Thanks for accepting the suggestion! Anything that involves wildcard iterator scares me, haha, so I'm doing extra scrutiny here :) 

 I have left few more comments that should be straight-forward to address.